### PR TITLE
Restored mixin based implementation of `observer` for class based components

### LINKED ISSUE #699: mobx-react 6 shallow testing regression
**What is the current behavior?**

 With `mobx-react` 6 upgrade regressions are detected in unit tests:

```jsx
// Foo.jsx
class Foo extends React.Component {
  //...
  render() {
    return <div>Test</div>;
  }
}
export default inject('store')(observer(Foo));
```
```jsx
//Foo.test.jsx
describe('Foo', () => {
  test('shallow', () => {
    const component = shallow(<Foo.wrappedComponent />);
    expect(component).toMatchSnapshot();
  });
})
```
Snapshot:
```jsx
<Observer>
  <Component />
</Observer>
```
Here's the repository with repro steps: https://github.com/sergei-startsev/mobx-react-6-regression

**What is the expected behavior?**

Expected snapshot:
```jsx
<div>
  Test
</div>
```
---
I believe the issue is caused by added `<Observer>` wrapper:
https://github.com/mobxjs/mobx-react/blob/4904b624f46be2e14d3a3d1017c1fa9bcb334f02/src/observer.js#L108

See https://github.com/mobxjs/mobx-react/commit/1ea40e21117d9c30067dd7b0267aa80b99473c59#diff-08f3e9ab539253c87ea111cc61c69e29R156.

### PR DESCRIPTION
**Pull request by mweststrate**
Proposal to revert to the classic implementation for class based observer components; by monkey patching the prototype rather than wrapping render in `<Observer>`

Fixes / affects #699, #692, #697, #702

### CODE REVIEW

File: src/observerClass.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import { PureComponent, Component } from "react"
+import { createAtom, _allowStateChanges, Reaction, $mobx } from "mobx"
+import { useStaticRendering as useStaticRenderingLite } from "mobx-react-lite"
+
+import { newSymbol, shallowEqual, setHiddenProp, patch } from "./utils/utils"
+
+let isUsingStaticRendering = false
```
Comment (FredyC) [inline src/observerClass.js]: Do you actually need to keep separate state for this? Why not reuse the variable from the Lite directly and export that setter directly as well?

Comment (mweststrate) [reply src/observerClass.js]: Because an imported variable is not bound to its original definition (that is, changes won't be reflected in it)

Comment (FredyC) [reply src/observerClass.js]: I don't think it should matter. The Lite exports both setter and getter. It does not export the actual variable. For consistency, the same should be exported here as well imo.

Comment (mweststrate) [reply src/observerClass.js]: ah missed that, will change!

Comment (mweststrate) [reply src/observerClass.js]: Done

**Comment by urugator** [comment]
Huh, haven't noticed that patching logic became so convoluted over time.
I know it's been mentioned many times that patching is necessary, but thb I am not sure about specifics. 
What's exactly problematic about inheriting the original component? Like:      
```javascript
class extends originalComponent {
  componentWillUnmount() {
    super();
    // dispose 
  }
}
```

**Comment by mweststrate** [comment]
Yeah, it's all disposeOnUnmount's fault :-(. It introduces a plethorea of
possiblities: the decorator want to set CWU, so does observer, so does the
component as defined by the user (either on prototype or by using an arrow
function). So the thing is monkey patches on different levels, which is
especially annoying if multiple inheritance trees are involved. See also
#705 that tries to simplifies things again, but breaks some cases

File: src/observerClass.js (hunk 1 above)
Comment (FredyC) [reply src/observerClass.js]: Nit: The observerClass should only consume `isUsingStaticRendering`, but both should be exported directly from `index.js` without extra wrapping.

File: src/observerClass.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    setHiddenProp(this, skipRenderKey, false)
+    /**
+     * forceUpdate will re-assign this.props. We don't want that to cause a loop,
+     * so detect these changes
+     */
+    setHiddenProp(this, isForcingUpdateKey, false)
+
+    // wire up reactive render
+    const baseRender = render.bind(this)
```
Comment (FredyC) [inline src/observerClass.js]: Perhaps move this line above `reactiveRender` declaration? It's fairly confusing to follow this code with all that hoisting. I would even move `const reaction` on top for a natural flow.

File: src/observerClass.js (hunk 1 above)
Comment (mweststrate) [reply src/observerClass.js]: Fixed

File: src/observerClass.js (hunk 2 above)
Comment (mweststrate) [reply src/observerClass.js]: yeah it is only the `disposeOnUnmount` decorator support that introduces the actual complexity, not the class support itself (which is pretty simple, and doesn't need the ugly component-might-not-have-been-committed hack :))

Comment (urugator) [reply src/observerClass.js]: > doesn't need the ugly component-might-not-have-been-committed hack

Are you sure? I think the reaction would have to be created in `componentDidMount` (similary to inside `useEffect`).
Check https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#adding-event-listeners-or-subscriptions:
Specifically:
> async rendering (where rendering might be interrupted before it completes, causing componentWillUnmount not to be called).

**Comment by mweststrate** [comment]
Dang, that is a good point. Would you mind opening an issue for that? IIRC
class components don't support this rendered-once but never comimtted
behavior. But in hindsight I'm not sure that actually makes sense.

**Comment by urugator** [comment]
Well technically there is a one already #645 :D

**Comment by mweststrate** [comment]
Fair point!

**Comment by urugator** [comment]
And I have another bad news ... this also affects `disposeOnUnmount`... actually `@disposeOnUnmount` is conceptually completely wrong, because constructors shouldn't contain side effects... all side effects should be in `cDM` or `cDU`.

**Comment by mweststrate** [comment]
Yeah I think we have to work to a hard split in the future: If you want to
use concurrent mode, limit yourself mobx-react-lite. If that is too hard of
a split, we could make v7 a minimal class implementation (as proposed
already) without all the other shizzle: Provider / inject / propTypes /
disposeOnUnmount

**Pull request merged**